Fixes Bug in Orientation of MaliputRailcar When it is Traveling Against s.

liangfok opened the pull request:
This change set was extracted from #5552. I believe it fixes a bug in `MaliputRailcar`'s orientation when it is traveling against `s`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5553)
<!-- Reviewable:end -->

liangfok commented:
+@maddog-tri for feature review since this modifies the return value of `maliput::api::Lane::GetOrientation()`.
+@jwnimmer-tri for car review please.

I'll assign a platform reviewer if / when this PR passes feature review since I'm not 100% sure this PR is correct.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-:-KfahtbGCwOVSNXkcG9D:b-xp5j7e)*
<!-- Sent from Reviewable.io -->

liangfok commented:
> this modifies the return value of `maliput::api::Lane::GetOrientation()`.

Sorry, I just realized that the above may be confusing. This PR does not modify the return value of the method. Instead, it takes the value that was returned and then modifies it to account for the fact that the `MaliputRailcar` might be facing in the direction that's opposite of increasing `s`.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-:-KfaiRYm9x-1HpiVM43Z:b-voyda8)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
:car: honk honk (assuming appropriate feature review checks the math).

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-matlab please.

maddog-tri commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput_railcar.cc, line 151 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KflMUrrU9_tZeI5o8s9:-KflMUrswx1iPhOXEfS5:b-2ymcg9) ([raw file](https://github.com/robotlocomotion/drake/blob/289ea29dfc7f9303fd7e44fe3ee6309e83a3b371/drake/automotive/maliput_railcar.cc#L151)):*
> ```cpp
>   const Rotation adjusted_rotation =
>       (lane_direction.with_s ? rotation :
>           Rotation(rotation.roll, rotation.pitch, rotation.yaw - M_PI));
> ```

This only works for flat roads, where `roll` and `pitch` are zero.  The correct way to reason about this is to begin with the complete 3d orientation of the car in the `(s,r,h)` frame, and then operate on that with the orientation of the lane-frame (with respect to world-frame, i.e., the output of `Lane::GetOrientation()`).

MaliputRailcar only has a 1-bit orientation in `(s,r,h)`:  it is either `RPY = (0,0,0)` or `RPY=(0,0,PI)`, so you can get away with doing the math on paper and skipping right to the two possible results in C++.  Either:
  * `(roll, pitch, yaw) ---> (roll, pitch, yaw)` for the "with s" case, OR
  * `(roll, pitch, yaw) ---> (-roll, -pitch, yaw - PI)` for the "against s" case.

(Also, with regards to dealing with `yaw - PI`:  I often keep angle arithmetic normalized by passing through `atan2()`.  The case here is not so exciting because the addend is a relatively small constant, but the result would be `(yaw - PI) == atan2(-sin(yaw), -cos(yaw))` in case you are so inclined.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput_railcar.cc, line 151 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KflMUrrU9_tZeI5o8s9:-KfmTyQlfx7PRN9SC42R:b-qqbim3) ([raw file](https://github.com/robotlocomotion/drake/blob/289ea29dfc7f9303fd7e44fe3ee6309e83a3b371/drake/automotive/maliput_railcar.cc#L151)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This only works for flat roads, where `roll` and `pitch` are zero.  The correct way to reason about this is to begin with the complete 3d orientation of the car in the `(s,r,h)` frame, and then operate on that with the orientation of the lane-frame (with respect to world-frame, i.e., the output of `Lane::GetOrientation()`).

MaliputRailcar only has a 1-bit orientation in `(s,r,h)`:  it is either `RPY = (0,0,0)` or `RPY=(0,0,PI)`, so you can get away with doing the math on paper and skipping right to the two possible results in C++.  Either:
  * `(roll, pitch, yaw) ---> (roll, pitch, yaw)` for the "with s" case, OR
  * `(roll, pitch, yaw) ---> (-roll, -pitch, yaw - PI)` for the "against s" case.

(Also, with regards to dealing with `yaw - PI`:  I often keep angle arithmetic normalized by passing through `atan2()`.  The case here is not so exciting because the addend is a relatively small constant, but the result would be `(yaw - PI) == atan2(-sin(yaw), -cos(yaw))` in case you are so inclined.)
</blockquote></details>

@maddog-tri, thanks for the review and thorough explanation of what _should_ be done. I fixed the code and added more unit test coverage. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 522 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnB0m6yF2uIRI9aL6v:-KfnB0m6yF2uIRI9aL6w:bad6smf) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L522)):*
> ```cpp
>   continuous_state()->set_speed(kInitialSpeed);
>   dut_->CalcOutput(*context_, output_.get());
>   const PoseVector<double>* pose2 = pose_output();
> ```

`pose2` is not symmetric with `against_s_pose`.  (`with_s_pose` would be more _apropose_, har-har.)

---

*[drake/automotive/test/maliput_railcar_test.cc, line 534 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnAmY2CvM7hOQW2Kne:-KfnAmY3MlwgUHLGjjLp:bq91fo6) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L534)):*
> ```cpp
>   // quaternions is equal to zero. The following tolerance was empirically
>   // determined.
>   EXPECT_NEAR(with_s_orientation.dot(against_s_orientation), 0, 1e-15);
> ```

Hmm... doesn't this yield the same result if `with_s_orientation == against_s_orientation`?

Perhaps a more direct comparison would also be more clear:  construct a quaternion for a 180deg yaw rotation, and multiply that by `with_s_orientation` (the order of operations matters of course).  The result should simply be equal to `against_s_orientation` --- and I think more clearly conveys the intended sense of "opposite".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 522 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnB0m6yF2uIRI9aL6v:-KfnCtW013MNSad2HZqJ:bg0010b) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L522)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

`pose2` is not symmetric with `against_s_pose`.  (`with_s_pose` would be more _apropose_, har-har.)
</blockquote></details>

(The variable name is what I am talking about, in case that wasn't clear.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 522 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnB0m6yF2uIRI9aL6v:-KfnE0K4q8xTyt0_wes6:b-n30ju3) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L522)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

(The variable name is what I am talking about, in case that wasn't clear.)
</blockquote></details>

Done. Har-har!

---

*[drake/automotive/test/maliput_railcar_test.cc, line 534 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnAmY2CvM7hOQW2Kne:-KfnE5bjzkW_10D_uYIZ:bp3zjs0) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L534)):*
> Hmm... doesn't this yield the same result if with_s_orientation == against_s_orientation?

If the two are equal, the dot product is 1.

I updated this PR to include your suggested test. Please let me know if I should remove the dot product test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 534 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnAmY2CvM7hOQW2Kne:-KfoU_gIpyDd8PGr2Omx:b-uuiib6) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L534)):*
> If the two are equal, the dot product is 1.

Yeah, my apologies for the dumb brain-spasm.

Well, I wouldn't keep both tests around (redundant).  Is the new test sensitive to the sign of `flip_yaw` (does -180 deg work, but not 180 deg?)?  If so, maybe the original test is a little more robust and a little less distracting -- in which case, I apologize for the noise.
Feel free to pick either one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-:-KfoYA88EVqw5bNa5jOy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@sammy-tri for platform review please.

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 534 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-KfnAmY2CvM7hOQW2Kne:-Kfpd3m3RebIcvZE4kDr:b-163mrn) ([raw file](https://github.com/robotlocomotion/drake/blob/004805d3b5699563be0fe3f914810d8b9d277862/drake/automotive/test/maliput_railcar_test.cc#L534)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

> If the two are equal, the dot product is 1.

Yeah, my apologies for the dumb brain-spasm.

Well, I wouldn't keep both tests around (redundant).  Is the new test sensitive to the sign of `flip_yaw` (does -180 deg work, but not 180 deg?)?  If so, maybe the original test is a little more robust and a little less distracting -- in which case, I apologize for the noise.
Feel free to pick either one.

</blockquote></details>

No problem! Yes, the new test is sensitive to whether the rotation is +180 or -180 degrees. I've thus removed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-:-Kfpdb4cfQNe3JZ0n2nk:b-1zm1j3)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental please.

maddog-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5553#-:-KfqO_lIRD175fuv48iT:bnfp4nl)*
<!-- Sent from Reviewable.io -->
